docs_src/design/adr/0023-North-South-Messaging.md

```diff
@@ -0,0 +1,158 @@
+# North-South Messaging
+
+## Status
+
+** Proposed **
+(as of 12/21/21)
+
+## Context and Proposed Design
+
+Today, data flowing from sensors/devices (the "southside") through EdgeX to enterprise applications, databases and cloud-based systems (the "northside") can be accomplished via REST or Message bus.  That is, sensor or device data collected by a device service can be sent via REST or message bus to core data.  Core data then relays the data to application services via message bus, but the sensor data can also be sent directly from device services to application services via message bus (by passing core data).  The message bus is implemented via Redis Pub/Sub (default) or via MQTT.  From the application services, data can be sent to northside endpoints in any number of ways – including via MQTT.
+
+So, in summary, data can be collected from a sensor or device and be sent from the southside to the northside entirely using message bus technology when desired.
+
+!!! Note
+    The message bus may also be implemented in Zero MQ but with some limitations.  As Zero MQ is being deprecated in EdgeX, it will not be considered further in this ADR.
+
+Communications from a 3rd party system (enterprise application, cloud application, etc.) to EdgeX in order to acuate a device or get the latest information from a sensor is accomplished via REST.  The 3rd party system makes a REST call of the command service which then relays a request to a device service also using REST.  There is no built in means to make a message-based request of EdgeX or the devices/sensors it manages.
+
+In a future release of EdgeX, there is a desire to allow 3rd party systems to make requests of the southside via message bus.  Specifically, a 3rd party system will send a command request via message to the command service via the allowed message bus implementations (which could be MQTT or Redis Pub/Sub today).  The command service would then relay the message request via message bus to the managing device service.  (The command service would also perform any translations on the request as it does for REST requests today.) The device service would use the message to trigger action on the device/sensor as it does when it receives a REST request today and respond via message bus back to the command service.  In turn, the command service would respond to the 3rd party system via message bus.
```

Comment:
@tonyespy , @jpwhitemn , IMO, we still need the Command Service to be the External MQTT connection which then publishes the command request to the specific Device service for the requested Device/Command. As with REST the Command service knows with Device service to route to publish to based on the Device specified in the request.

Example:
   External topic received is : `edgex/command/request/<device>/<command-name>/...`
   Command Service forwards the command on the internal Edgex MessageBus with the topic: `edgex/command/request/<device-service>/<device>/<command-name>/...`

Alternatively we could have the Device Services subscribe to `edgex/command/request/<device>/#` for each device when they are added. DS would also have to unsubscribe when devices are removed. This would allow any EdgeX service to send a command via the internal EdgeX MessageBus using the topic `edgex/command/request/<device>/<command-name>/...`

Reply:
> Also as others (@cloudxxx8, @lenny-intel) have pointed out, for this to work we'd need to grant these 3rd party applications access to our message bus, which has already been a bit problematic with Kuiper. Since our default message bus impl is Redis pub/sub, this means we would need some kind of mechanism to securely distribute the Redis credentials (e.g. user/password) to these external applications.
> 
> It also should be noted that this approach would completely bypass the API Gateway.
> 
> If on the other hand, we secure our MQTT message bus implementation, authentication/authorization of external clients could be handled via client TLS client authentication, which plays into our discussions about Spiffe/Spire (see the [Delayed Services ADR](https://github.com/edgexfoundry/edgex-docs/blob/main/docs_src/design/adr/security/0020-spiffe.md) and/or [Microservices Auth ADR](https://github.com/edgexfoundry/edgex-docs/pull/659)).
> 
> Moving to a model where we instead use an external MQTT broker for this purpose means we've pushed all responsibility for securing communications to EdgeX to the entity operating this broker.

From a security perspective, there are two separate problems:

1. Granting parties the authorization to talk on the message bus.  Pretty much every MQTT broker has a way to do this, where the most common are username/password and certificate-based mutual-auth TLS.  SPIFFE/SPIRE is a good way to distribute such certificates if one can manage the requirements for certificate rotation on a continual basis.  Otherwise, username/password will probably end up being used (with server-side TLS if the broker is off-box)
2. Enabling the receiving microservice to positively identity the sending microservice.  It is an uncommon message broker feature for the message broker to pass metadata as to the sender of a message.  Usually, however, we have to roll our own, such as putting an access_token field in the message body that has a JWT, for example.